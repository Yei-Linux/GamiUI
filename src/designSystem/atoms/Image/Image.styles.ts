import styled from '@emotion/styled'

export const Img = styled.img<{
  maxWidth: string
  maxHeight: string
  minWidth: string
  minHeight: string
  width: string
  height: string
}>`
  width: ${(props) => props.width};
  height: ${(props) => props.height};
  max-width: ${(props) => props.maxWidth};
  min-width: ${(props) => props.minWidth};
  max-height: ${(props) => props.maxHeight};
  min-height: ${(props) => props.minHeight};
`

export const BackgroundImg = styled.div<{
  src: string
  maxWidth: string
  maxHeight: string
  minWidth: string
  minHeight: string
  width: string
  height: string
}>`
  background-image: url(${(props) => props.src});
  background-size: contain;
  background-repeat: no-repeat;
  background-position: center center;
  width: ${(props) => props.width};
  height: ${(props) => props.height};
  max-width: ${(props) => props.maxWidth};
  min-width: ${(props) => props.minWidth};
  max-height: ${(props) => props.maxHeight};
  min-height: ${(props) => props.minHeight};
`
