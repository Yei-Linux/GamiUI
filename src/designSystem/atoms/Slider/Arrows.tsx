import { TOnClick } from 'core/domain/types'
import KeenSlider from 'keen-slider'
import React, { Fragment } from 'react'

import * as S from './Slider.styles'

export interface ISliderElements {
  slider: KeenSlider
  vertical: boolean
  currentSlide: number
}

const Arrows = ({ slider, vertical, currentSlide }: ISliderElements) => {
  const isFirstSlide = () => currentSlide == 0

  const isLastSlide = () => currentSlide == slider.details().size - 1

  const handleClickPrev = (e: TOnClick) => {
    e.stopPropagation()
    slider.prev()
  }

  const handleClickNext = (e: TOnClick) => {
    e.stopPropagation()
    slider.next()
  }

  return (
    <Fragment>
      {slider && !vertical && (
        <>
          <S.ArrowLeft
            onClick={handleClickPrev}
            size="25px"
            name={isFirstSlide() ? 'arrow__left__semiwhite' : 'arrow__left'}
          />
          <S.ArrowRight
            size="25px"
            onClick={handleClickNext}
            name={isLastSlide() ? 'arrow__right__semiwhite' : 'arrow__right'}
          />
        </>
      )}
    </Fragment>
  )
}

export default Arrows
