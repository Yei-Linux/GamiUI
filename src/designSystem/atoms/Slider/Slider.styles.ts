import { css } from '@emotion/react'
import styled from '@emotion/styled'
import { colorLight } from '../../../styles/theme'
import Icon from '../Icon'

export const SliderWrapper = styled.div`
  position: relative;
`

export const KeenSlider = styled.div<{ isVertical?: boolean }>`
  display: flex;
  overflow: hidden;
  position: relative;
  user-select: none;
  -webkit-touch-callout: none;
  -khtml-user-select: none;
  -ms-touch-action: pan-y;
  touch-action: pan-y;
  -webkit-tap-highlight-color: transparent;

  ${(props) =>
    props.isVertical &&
    css`
      flex-wrap: wrap;
      height: 300px;
    `}

  .keenslider__slide {
    min-width: calc(100% - 0px);
    max-width: calc(100% - 0px);
    transform: translate3d(0px, 0px, 0px);
    width: 100%;
    min-height: 100%;
    overflow: hidden;
    position: relative;
  }
`

export const Dots = styled.div`
  display: flex;
  padding: 10px 0;
  justify-content: center;
`

export const Arrow = css`
  width: 30px;
  height: 30px;
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  -webkit-transform: translateY(-50%);
  cursor: pointer;
`

export const ArrowLeft = styled(Icon)`
  ${Arrow}
  left: 5px;
`

export const ArrowRight = styled(Icon)`
  ${Arrow}
  left: auto;
  right: 5px;
`

export const Dot = styled.button<{ isActive: boolean }>`
  border: none;
  width: 10px;
  height: 10px;
  background: #c5c5c5;
  border-radius: 50%;
  margin: 0 5px;
  padding: 5px;
  cursor: pointer;

  &:focus {
    outline: none;
  }

  ${(props) =>
    props.isActive &&
    css`
      background: ${colorLight.primary.one};
    `}
`
