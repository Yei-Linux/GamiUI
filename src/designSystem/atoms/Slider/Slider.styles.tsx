import { css } from '@emotion/react'
import styled from '@emotion/styled'
import { mixinFlexVariants } from 'styles/mixins/flex'
import { theme } from 'styles/tokens'
import Icon from '../Icon'

export const Slider = styled.div`
  position: relative;
`

export const KeenSliderLib = styled.div`
  display: flex;
  position: relative;
  overflow: hidden;

  user-select: none;
  -webkit-touch-callout: none;
  -khtml-user-select: none;
  -ms-touch-action: pan-y;
  touch-action: pan-y;
  -webkit-tap-highlight-color: transparent;

  &.vertical {
    flex-wrap: wrap;
    height: 450px;
  }

  &.keenslider__slide {
    min-width: calc(100% - 0px);
    max-width: calc(100% - 0px);
    transform: translate3d(0px, 0px, 0px);
    width: 100%;
    min-height: 100%;
  }
`

export const Arrow = css`
  width: 30px;
  height: 30px;

  position: absolute;
  top: 50%;

  transform: translateY(-50%);

  &:hover {
    cursor: pointer;
  }
`

export const ArrowLeft = styled(Icon)`
  ${Arrow}
  left: 5px;
`

export const ArrowRight = styled(Icon)`
  ${Arrow}
  left: auto;
  right: 5px;
`

export const Dots = styled.div`
  ${mixinFlexVariants({ justifyContent: 'center' })}
  padding: 10px 0;
`

export const Dot = styled.button`
  border: none;
  border-radius: 50%;
  margin: 0 5px;
  padding: 5px;

  width: 10px;
  height: 10px;
  background: ${theme.light.neutral[500]};

  &:focus {
    outline: none;
  }

  &:hover {
    cursor: pointer;
  }

  &.active {
    background: ${theme.light.primary.jordyBlue};
  }
`
