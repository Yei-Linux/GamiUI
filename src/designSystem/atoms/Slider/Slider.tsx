import React, { Fragment, useEffect, useRef, useState } from 'react'
import { useKeenSlider } from 'keen-slider/react'
import {
  ArrowLeft,
  ArrowRight,
  Dot,
  Dots,
  KeenSlider,
  SliderWrapper,
} from './Slider.styles'
import { colorLight } from '../../../styles/theme'
import useDevice from '../../../hooks/useDevice'

export interface SliderProps {
  children: React.ReactNode
  slidesPerView: ISliderPerView
  spacing: number
  initial: number
  isLoop: boolean
  vertical?: boolean
  isAutoplay?: boolean
}

interface ISliderPerView {
  phone: number
  tablet: number
  laptop: number
  desktop: number
}

const Slider = ({
  children,
  slidesPerView,
  spacing,
  initial,
  isLoop,
  vertical,
  isAutoplay,
}: SliderProps) => {
  const { device } = useDevice()
  const [pause, setPause] = useState(false)
  const timer: any = useRef()
  const [currentSlide, setCurrentSlide] = useState(0)
  const [sliderRef, slider]: any = useKeenSlider({
    initial,
    loop: isLoop,
    slidesPerView: device ? slidesPerView?.[device] : 1,
    spacing,

    vertical,
    centered: vertical,

    duration: 1000,
    dragStart: () => {
      setPause(true)
    },
    dragEnd: () => {
      setPause(false)
    },

    slideChanged(s) {
      setCurrentSlide(s.details().relativeSlide)
    },
  })

  const handleAutoplay = () => {
    sliderRef?.current?.addEventListener('mouseover', () => {
      setPause(true)
    })
    sliderRef?.current?.addEventListener('mouseout', () => {
      setPause(false)
    })
  }

  const handleIntervalOnAutoplay = () => {
    timer.current = setInterval(() => {
      if (!pause && slider) {
        slider.next()
      }
    }, 2000)
    return () => {
      clearInterval(timer.current)
    }
  }

  useEffect(() => {
    isAutoplay && handleAutoplay()
  }, [sliderRef])

  useEffect(() => {
    if (isAutoplay) return handleIntervalOnAutoplay()
  }, [pause, slider])

  return (
    <Fragment>
      <SliderWrapper className="slider__wrapper">
        <KeenSlider
          isVertical={vertical}
          ref={sliderRef}
          className="keen-slider"
        >
          {children}
        </KeenSlider>
        {slider && !vertical && (
          <>
            <ArrowLeft
              onClick={(e: any) => e.stopPropagation() || slider.prev()}
              fill={
                currentSlide === 0
                  ? colorLight.neutral.five
                  : colorLight.primary.one
              }
              name="arrowLeft"
            />
            <ArrowRight
              onClick={(e: any) => e.stopPropagation() || slider.next()}
              fill={
                currentSlide === slider.details().size - 1
                  ? colorLight.neutral.five
                  : colorLight.primary.one
              }
              name="arrowRight"
            />
          </>
        )}
      </SliderWrapper>

      {slider && !vertical && (
        <Dots className="slider__dots__wrapper">
          {[...Array(slider.details().size).keys()].map((idx) => (
            <Dot
              key={idx}
              onClick={() => {
                slider.moveToSlideRelative(idx)
              }}
              isActive={currentSlide === idx}
            />
          ))}
        </Dots>
      )}
    </Fragment>
  )
}

Slider.defaultProps = {
  initial: 0,
  isLoop: true,
  slidesPerView: {
    phone: 1,
    tablet: 3,
    laptop: 5,
    desktop: 6,
  },
  spacing: 0,
  vertical: false,
  isAutoplay: false,
}

export default Slider
