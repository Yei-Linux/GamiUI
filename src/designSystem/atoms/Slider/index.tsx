export { default } from './Slider'
export { options } from './constants'
