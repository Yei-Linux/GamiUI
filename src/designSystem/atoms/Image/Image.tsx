import React, { Fragment } from 'react'
import { BackgroundImg, Img } from './Image.styles'

interface ImageLink {
  url?: string
  newTab?: boolean
}

export interface ImageProps {
  children?: React.ReactNode
  src: string
  alt?: string
  maxWidth: string
  maxHeight: string
  minWidth: string
  minHeight: string
  width: string
  height: string
  link?: ImageLink | null
}

const Image = ({
  children,
  src,
  alt,
  maxWidth,
  maxHeight,
  minWidth,
  minHeight,
  width,
  height,
  link,
}: ImageProps) => {
  const imageCommonsProps = {
    src,
    maxWidth,
    maxHeight,
    minWidth,
    minHeight,
    width,
    height,
  }
  const shouldOpenLinkInNewTab = link?.newTab
  const imageElement = children ? (
    <BackgroundImg {...imageCommonsProps}>{children}</BackgroundImg>
  ) : (
    <Img {...imageCommonsProps} alt={alt} />
  )
  const maybeLink = link ? (
    <a
      href={link?.url}
      target={shouldOpenLinkInNewTab ? '_blank' : undefined}
      rel="noreferrer"
    >
      {imageElement}
    </a>
  ) : (
    <Fragment>{imageElement}</Fragment>
  )

  return <Fragment>{maybeLink}</Fragment>
}

Image.defaultProps = {
  children: null,
  alt: '',
  link: null,
  width: 'auto',
  height: 'auto',
}

export default Image
