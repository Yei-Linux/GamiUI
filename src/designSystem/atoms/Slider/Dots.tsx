import classNames from 'classnames'
import React, { Fragment } from 'react'
import { ISliderElements } from './Arrows'
import * as S from './Slider.styles'

const Dots = ({ slider, vertical, currentSlide }: ISliderElements) => {
  return (
    <Fragment>
      {slider && !vertical && (
        <S.Dots className={classNames('slider__dots__wrapper')}>
          {[...Array(slider.details().size).keys()].map((idx) => (
            <S.Dot
              key={idx}
              onClick={() => {
                slider.moveToSlideRelative(idx)
              }}
              className={classNames({
                active: currentSlide === idx,
              })}
            />
          ))}
        </S.Dots>
      )}
    </Fragment>
  )
}

export default Dots
