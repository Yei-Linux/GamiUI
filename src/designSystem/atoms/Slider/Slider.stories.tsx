import React from 'react'
import { Meta } from '@storybook/react'

import Slider from '.'
import Image from '../Image'

export default {
  title: 'Atoms/Slider',
  component: Slider,
  args: { ...Slider.defaultProps },
  argTypes: {},
} as Meta

const Children = () => (
  <>
    <Image
      width="100%"
      height="350px"
      className="keen-slider__slide"
      src="https://pixnio.com/free-images/2017/03/01/2017-03-01-07-11-56.jpg"
    />
    <Image
      width="100%"
      height="350px"
      className="keen-slider__slide"
      src="https://pixnio.com/free-images/2017/03/01/2017-03-01-07-11-56.jpg"
    />
    <Image
      width="100%"
      height="350px"
      className="keen-slider__slide"
      src="https://pixnio.com/free-images/2017/03/01/2017-03-01-07-11-56.jpg"
    />
    <Image
      width="100%"
      height="350px"
      className="keen-slider__slide"
      src="https://pixnio.com/free-images/2017/03/01/2017-03-01-07-11-56.jpg"
    />
    <Image
      width="100%"
      height="350px"
      className="keen-slider__slide"
      src="https://pixnio.com/free-images/2017/03/01/2017-03-01-07-11-56.jpg"
    />
    <Image
      width="100%"
      height="350px"
      className="keen-slider__slide"
      src="https://pixnio.com/free-images/2017/03/01/2017-03-01-07-11-56.jpg"
    />
  </>
)

export const Basic = (args: any) => {
  return (
    <Slider {...args}>
      <Children />
    </Slider>
  )
}
Basic.args = {}

export const Autoplay = (args: any) => {
  return (
    <Slider {...args}>
      <Children />
    </Slider>
  )
}
Autoplay.args = {
  isAutoplay: true,
}

export const Vertical = (args: any) => {
  return (
    <Slider {...args}>
      <Children />
    </Slider>
  )
}
Vertical.args = {
  vertical: true,
}
