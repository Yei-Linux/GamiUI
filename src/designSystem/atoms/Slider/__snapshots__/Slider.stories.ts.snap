// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Atoms/Slider Autoplay 1`] = `
<div
  className="css-g1apgy-SliderWrapper etxlari0"
>
  <div
    className="keen-slider css-xuu3ag-KeenSlider etxlari1"
  >
    <img
      alt=""
      className="keen-slider__slide css-pkaxp9-Img efl6h300"
      height="350px"
      src="https://pixnio.com/free-images/2017/03/01/2017-03-01-07-11-56.jpg"
      width="100%"
    />
    <img
      alt=""
      className="keen-slider__slide css-pkaxp9-Img efl6h300"
      height="350px"
      src="https://pixnio.com/free-images/2017/03/01/2017-03-01-07-11-56.jpg"
      width="100%"
    />
    <img
      alt=""
      className="keen-slider__slide css-pkaxp9-Img efl6h300"
      height="350px"
      src="https://pixnio.com/free-images/2017/03/01/2017-03-01-07-11-56.jpg"
      width="100%"
    />
    <img
      alt=""
      className="keen-slider__slide css-pkaxp9-Img efl6h300"
      height="350px"
      src="https://pixnio.com/free-images/2017/03/01/2017-03-01-07-11-56.jpg"
      width="100%"
    />
    <img
      alt=""
      className="keen-slider__slide css-pkaxp9-Img efl6h300"
      height="350px"
      src="https://pixnio.com/free-images/2017/03/01/2017-03-01-07-11-56.jpg"
      width="100%"
    />
    <img
      alt=""
      className="keen-slider__slide css-pkaxp9-Img efl6h300"
      height="350px"
      src="https://pixnio.com/free-images/2017/03/01/2017-03-01-07-11-56.jpg"
      width="100%"
    />
  </div>
</div>
`;

exports[`Storyshots Atoms/Slider Basic 1`] = `
<div
  className="css-g1apgy-SliderWrapper etxlari0"
>
  <div
    className="keen-slider css-xuu3ag-KeenSlider etxlari1"
  >
    <img
      alt=""
      className="keen-slider__slide css-pkaxp9-Img efl6h300"
      height="350px"
      src="https://pixnio.com/free-images/2017/03/01/2017-03-01-07-11-56.jpg"
      width="100%"
    />
    <img
      alt=""
      className="keen-slider__slide css-pkaxp9-Img efl6h300"
      height="350px"
      src="https://pixnio.com/free-images/2017/03/01/2017-03-01-07-11-56.jpg"
      width="100%"
    />
    <img
      alt=""
      className="keen-slider__slide css-pkaxp9-Img efl6h300"
      height="350px"
      src="https://pixnio.com/free-images/2017/03/01/2017-03-01-07-11-56.jpg"
      width="100%"
    />
    <img
      alt=""
      className="keen-slider__slide css-pkaxp9-Img efl6h300"
      height="350px"
      src="https://pixnio.com/free-images/2017/03/01/2017-03-01-07-11-56.jpg"
      width="100%"
    />
    <img
      alt=""
      className="keen-slider__slide css-pkaxp9-Img efl6h300"
      height="350px"
      src="https://pixnio.com/free-images/2017/03/01/2017-03-01-07-11-56.jpg"
      width="100%"
    />
    <img
      alt=""
      className="keen-slider__slide css-pkaxp9-Img efl6h300"
      height="350px"
      src="https://pixnio.com/free-images/2017/03/01/2017-03-01-07-11-56.jpg"
      width="100%"
    />
  </div>
</div>
`;

exports[`Storyshots Atoms/Slider Vertical 1`] = `
<div
  className="css-g1apgy-SliderWrapper etxlari0"
>
  <div
    className="keen-slider css-ullxii-KeenSlider etxlari1"
  >
    <img
      alt=""
      className="keen-slider__slide css-pkaxp9-Img efl6h300"
      height="350px"
      src="https://pixnio.com/free-images/2017/03/01/2017-03-01-07-11-56.jpg"
      width="100%"
    />
    <img
      alt=""
      className="keen-slider__slide css-pkaxp9-Img efl6h300"
      height="350px"
      src="https://pixnio.com/free-images/2017/03/01/2017-03-01-07-11-56.jpg"
      width="100%"
    />
    <img
      alt=""
      className="keen-slider__slide css-pkaxp9-Img efl6h300"
      height="350px"
      src="https://pixnio.com/free-images/2017/03/01/2017-03-01-07-11-56.jpg"
      width="100%"
    />
    <img
      alt=""
      className="keen-slider__slide css-pkaxp9-Img efl6h300"
      height="350px"
      src="https://pixnio.com/free-images/2017/03/01/2017-03-01-07-11-56.jpg"
      width="100%"
    />
    <img
      alt=""
      className="keen-slider__slide css-pkaxp9-Img efl6h300"
      height="350px"
      src="https://pixnio.com/free-images/2017/03/01/2017-03-01-07-11-56.jpg"
      width="100%"
    />
    <img
      alt=""
      className="keen-slider__slide css-pkaxp9-Img efl6h300"
      height="350px"
      src="https://pixnio.com/free-images/2017/03/01/2017-03-01-07-11-56.jpg"
      width="100%"
    />
  </div>
</div>
`;
